### ISSUE TITLE
Introduce CompositionAbstract for Mandatory fields, Language, Subject, TimeValue, Origin Value etc. 

### ISSUE THREAD

**Issue by SevKohler** [ehrbase/fhir-bridge#93]
## Background ##

We have a lot of duplicates when it comes down to the ¨Mandatory fields¨ e.g. Language, Territory, StartTimeValue etc. and things like this. This problem does appear on different level of the mapping also within Observations contained in a Composition etc. .
I suggest creating an Abstract Composition and Observation etc. where we have a method to set these Values (```setMandatoryFields()```), the from the opt generated classes then only needs to inherit these classes. This would also thin out our 10000 duplicate [save methods](https://github.com/ehrbase/fhir-bridge/blob/develop/src/main/java/org/ehrbase/fhirbridge/rest/EhrbaseService.java), since we could do one ```save(CompositionAbstract composition)``` and use it for every Composition type. I already mentioned this in the openEHR_SDK as an [issue](https://github.com/ehrbase/openEHR_SDK/issues/94), but since im running into this nonstop its maybe easier to do the extends by hand for now. If not we will loads and loads of duplicated code and if we want to change smt. like retrieving the Language we want to set for the for the Composition from the Template (as it was mentioned in a comment), we need to change this at 100 different places. 

I implemented this, for now only for the bloodgasmapping: [abstract](https://github.com/ehrbase/fhir-bridge/blob/feature/60_mapping_blood_gas/src/main/java/org/ehrbase/fhirbridge/opt/CompositionAbstract.java), [saveMethod](https://github.com/ehrbase/fhir-bridge/blob/2e07eb03b74b1d280f277fd65758eb7bf664de82/src/main/java/org/ehrbase/fhirbridge/rest/EhrbaseService.java#L174). I would recommend to use these classes for future mappings and alter them if necessary.
## Acceptance criteria ##

_These are the functional requirements for this story. Add all steps to be implemented to fulfill the requirements._

- [ ] Item 1

## Success criteria ##

- [ ] Review / Merge request approved (P2P session)
- [ ] Unit tests passed
- [ ] Updated documentation (Javadoc and Sphinx)
- [ ] Acceptance criteria fulfilled
- [ ] Build without errors
- [ ] Release notes prepared
- [ ] Runtime warnings

**Comment by SevKohler** [comment]
@ppazos @subigre

**Comment by ppazos** [comment]
@SevKohler I agree with this request, I even mentioned this months ago on our planning meetings when I started to work with the bridge.

1. The issue should not be here but in the SDK repo as you already raised it there.
2. When that is fixed in the SDK generator, a refactoring should be made on the bridge, since all OPT classes would need to be regenerated and some other classes will be affected.
3. That refactoring ticket should be here depending on the ticket in the SDK repo.

**Comment by SevKohler** [comment]
> 
> 
> @SevKohler I agree with this request, I even mentioned this months ago on our planning meetings when I started to work with the bridge.
> 
>     1. The issue should not be here but in the SDK repo as you already raised it there.
> 
>     2. When that is fixed in the SDK generator, a refactoring should be made on the bridge, since all OPT classes would need to be regenerated and some other classes will be affected.
> 
>     3. That refactoring ticket should be here depending on the ticket in the SDK repo.

We could already use the placeholder classes as i supposed and later on replcae them with the ones from the SDK that would safe a lot of refactoring time instead of writing loads and lodas of duplicates. There is no harm in doing so, this only safes time.

**Comment by ppazos** [comment]
@SevKohler there are a couple of classes growing because of not having an abstract superclass for all COMPOSITOINS, but it's just a couple of lines each time and are the same classes. I don't think this will cause issues for a future refactoring, since everything is encapsulated and we know which classes to modify.

On the other side, if we modify the code of generated classes manually, and we need to update an OPT (OPTs change), we will need to generate and modify the code again, which is more costly than just fix the SDK generator once and do the refactoring on the FHIR bridge for adding the abstract COMPOSITION for all generated Composition* classes.

**Comment by SevKohler** [comment]
> 
> 
> @SevKohler there are a couple of classes growing because of not having an abstract superclass for all COMPOSITOINS, but it's just a couple of lines each time and are the same classes. I don't think this will cause issues for a future refactoring, since everything is encapsulated and we know which classes to modify.
> 
> On the other side, if we modify the code of generated classes manually, and we need to update an OPT (OPTs change), we will need to generate and modify the code again, which is more costly than just fix the SDK generator once and do the refactoring on the FHIR bridge for adding the abstract COMPOSITION for all generated Composition* classes.

I see your point, nevertheless adding a small ```extends CompositionAbstract``` to each composition is not a lot alteration that needs to be done for each opt update that happens from time to time. The other solution will generate more and more code that needs to refactored and cost more time then the small extends when updating an opt. I thing this is a bad idea and only pushes the problem to a point where its harder to be refactored. Nevertheless for now i will do this for my mappings and take care for the opt updates and refactoring of them later on.

**Comment by ppazos** [comment]
@SevKohler I'm not arguing on how big or small changes are. My point is about maintainability. Any code that is auto generated shouldn't be manually modified because any regeneration of code will break the manual changes and will require to do the manual work each time something auto-generated is updated.

Another point would be: this process will require everyone adding new mappings to modify their auto-generated code manually. I don't think adding an extra step to the process, when that could be automated, is really worth it.

I'm in favor of pushing the ticket in the SDK repo, I actually mentioned it to @birgerhaarbrandt, and closing this one.

**Comment by SevKohler** [comment]
Well i did it like that for both mappings both had opt changes and i found it way more pleasant and easy then duplicating, but since your pushing the ticket problem is solved any way!

**Comment by subigre** [comment]
I apologize because I missed this issue.
In my case I don't need an abstract class but, as discussed with @ppazos, I would appreciate if we could introduce an interface for all composition classes generated by the SDK.
This could help in the context of the PoC and we might be able to define a CompositionConverter to have a common interface in order to implement the conversion logic between composition and FHIR resources (like Spring defines MessageConverter for JMS).

**Comment by SevKohler** [comment]
I think an AbstractComposition is better, this is also because of the semantics itself and the Mandatory(
```Java
observation.setOriginValue(effectiveDateTime);
observation.setTimeValue(effectiveDateTime);
observation.setLanguage(Language.DE);
```) 
Fields we always have to set, using an abstract class for that is better in design and use. 
Also implementing here to get the language from the Template would eradicate a lot of duplicates.
An Interface is to  be used when classes have the same methods but differ in their execution, we have here several cases where we need the same method for subclasses.

**Comment by SevKohler** [comment]
The Converter then contains these mandatory fields and sets them for each Composition ?

**Comment by subigre** [comment]
The CompositionConverter is an generic interface.
```
package org.ehrbase.fhirbridge.camel.component.ehr.composition;

import org.ehrbase.fhirbridge.ehr.Composition;

public interface CompositionConverter<C extends Composition, T> {

    T fromComposition(C composition);

    C toComposition(T object);
}
```

But we might have an `AbstractCompositionConverter` that implements the common logic

**Comment by SevKohler** [comment]
Okay !
